In [1]:
import numpy as np
from numpy import linalg
from scipy.linalg import expm

1. Until T = 100, simulate a continuous-time Markov chain with generator matrix Q, starting from state 1. Find the stationary distribution by calculating the share of time spent in each state. Compare with the theoretical value.

In [2]:
np.random.seed(2)  # Seeds the generator
Q = np.array([[-3,1,2], [2,-5,3], [5,0,-5]])  # Generator matrix

pi = {s:0 for s in range(3)}  # Dictionary of share of time spent in each state
T = 0 
state = 0
while T < 100:    
    timesT = {s:0 for s in range(3)}  
    for i in range(3):
        if Q[state][i] <= 0:
            timesT[i] = math.inf
        else:
            timesT[i] = np.random.exponential(1/(Q[state][i]))  # Simulated Exponential Clocks
    interTime = min(timesT.values())  # Time spent in current state before moving to next state
    
    for i in range(3):  # Searches for which state does the interTime belongs to
        if timesT[i] == interTime:  
            pi[state] += interTime  # Sum of time spent in current state
            state = i  # Next state
            break
    
    T += interTime

for i in range(3):
    pi[i] /= 100

print(list(pi.values()))  # Stationary Distribution

[0.5792756231961823, 0.11657500510271701, 0.31010426871398394]


In [3]:
# Theoretical value of Stationary Distribution until T = 100
list(expm(100*Q)[1]) 

[0.581395348837152, 0.11627906976743044, 0.3023255813953194]

Alternatively, <br/>
$-3\pi_{1} + 2\pi_{2} + 5\pi_{3} = 0$ <br/>
$\pi_{1} - 5\pi_{2} = 0$ <br/>
$2\pi_{1} + 3\pi_{2} - 5\pi_{3} = 0$ <br/>
$\pi_{1} + \pi_{2} + \pi_{3} = 1$. <br/>
Solving the system of linear equations, we get $\pi_{1} = \frac{25}{43} \approx 0.581395$, $\pi_{2} = \frac{5}{43} \approx 0.116279$, and $\pi_{3} = \frac{13}{43} \approx 0.302325$. 

2. Simulate 1000 times, starting from state 2, a continuous-time Markov chain with generator matrix, Q. Find the empirical distribution of this chain at time t = 2. Compare with the theoretical value.

In [4]:
np.random.seed(2) # Seed the generator 
Q = np.array([[-3,3],[2,-2]]) # Generator Matrix

proportions = {s:0 for s in range(2)}  # Dictionary of number of times spent in each state

for k in range(1000):
    T = 0 
    n = 1 # Starts at state 2
    while T < 2:
        n = state 
        timesT = {s:0 for s in range(2)}  
        for i in range(2):
            if Q[n][i] <= 0:
                timesT[i] = math.inf
            else:
                timesT[i] = np.random.exponential(1/(Q[state][i]))  # Simulated Exponential Clocks
        interTime = min(timesT.values())  # Time spent in current state before moving to next state

        for j in range(2):  # Searches for which state does the interTime belongs to
            if timesT[j] == interTime:  
                state = j  # Next state
                break
        T += interTime
    proportions[n] += 1 # Adds up the count of which state is at t=2
            
for i in range(2):
    proportions[i] /= 1000

print(list(proportions.values()))   

[0.402, 0.598]


Computation of Theoretical Value:  

First, solve for stationary distribution. $\pi A = 0$.  $-3\pi_{1} + 2\pi_{2} = 0$ implies $\pi_{1} = \frac{2}{3}\pi_{2}$  

Plugging into $\begin{eqnarray} \pi_{1}+\pi_{2}=1\end{eqnarray}$ gives us $\pi_{1} = \frac{2}{5}$ and $\pi_{2} = \frac{3}{5}$, so $\pi = [\frac{2}{5} \;\; \frac{3}{5}] $. 


Then we solve for the eigenvalues of Q: det$(A-\lambda I_{2}) = 0$ so $\lambda^{2} + 5\lambda = 0$ which gives us $\lambda_{1,2} = 0, -5$. Afterwards, we find the eigenvectors corresponding to $\lambda_{2} = -5$ which is $u = [1 \;\;\; -1]$ since $\lambda_{1} = \pi$. Next, we find $x(0) = c_{1}\pi + c_{2}u$ which gives us $c_{1} = 1$ and $c_{2} = \frac{-2}{5}$.

Therefore, $x_{1}(2) = \frac{2}{5} + \frac{-2}{5}e^{-10} \approx 0.39998$ and $x_{2}(2) = \frac{3}{5} + \frac{-2}{5}e^{-10} \approx 0.59998$. The rate of convergence is $e^{-10}$ since $t=2$. 